File: whitebox-attacks-assessment.py
import hashlib
import sys

def pw_hash(password):
    salt = 'it6z'
    hash_object = hashlib.md5((salt + password).encode())
    return hash_object.hexdigest()

def main():
    try:
        number = 0
        while True:
            password = str(number)
            hashed_value = pw_hash(password)
            # Check if hash starts with '0e' and has numbers after 'e'
            if hashed_value.startswith('0e') and hashed_value[2:].isdigit():
                print(f"Found hash starting with '0e', followed by numbers:")
                print(f"Password: {password}")
                print(f"Hash: {hashed_value}")
                break
            number += 1
    except Exception as e:
        print(f"An error occurred: {e}")

if __name__ == "__main__":
    main()

File: whitebox-attacks-assessment.c
#include <stdio.h>
#include <string.h>
#include <openssl/md5.h>

// Function to calculate MD5 hash
void pw_hash(const char *password, char *output) {
    char salt[] = "it6z";
    char input[256];
    unsigned char hash[MD5_DIGEST_LENGTH];

    strcpy(input, salt);
    strcat(input, password);
    
    MD5((unsigned char *)input, strlen(input), hash);

    for (int i = 0; i < MD5_DIGEST_LENGTH; i++) {
        sprintf(output + (i * 2), "%02x", hash[i]);
    }
    output[MD5_DIGEST_LENGTH * 2] = '\0';
}

int main() {
    int number = 0;
    char password[256];
    char hashed_value[MD5_DIGEST_LENGTH * 2 + 1];

    while (1) {
        sprintf(password, "%d", number);
        pw_hash(password, hashed_value);
        
        // Check if hash starts with '0e' and has digits after 'e'
        if (strncmp(hashed_value, "0e", 2) == 0) {
            int is_digit = 1;
            for (int i = 2; i < MD5_DIGEST_LENGTH * 2; i++) {
                if (!isdigit(hashed_value[i])) {
                    is_digit = 0;
                    break;
                }
            }
            if (is_digit) {
                printf("Found hash starting with '0e', followed by numbers:\n");
                printf("Password: %s\n", password);
                printf("Hash: %s\n", hashed_value);
                break;
            }
        }
        number++;
    }

    return 0;
}
